spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala

```diff
@@ -0,0 +1,204 @@
+          matchedConditions = actions.map(getClauseCondition),
+          matchedOutputs = actions.map(actionOutput(_, targetOutputCols)),
+          notMatchedConditions = notActions.map(getClauseCondition),
+          notMatchedOutputs = notActions.map(actionOutput(_, targetOutputCols)),
+          targetOutput = targetOutputCols :+ Literal(false),
+          deleteOutput = targetOutputCols :+ Literal(true),
+          joinedAttributes = joinPlan.output
+        )
+        val joinKeysFromTarget = targetOutputCols.filter (
+          attr => cond.references.exists(attr.semanticEquals(_))
+        )
+        val writePlan = buildWritePlan(joinPlan, Project(target.output, joinPlan), target.table, joinKeysFromTarget)
+        val mergePlan = MergeInto(mergeParams, target, writePlan)
+        val batchWrite = mergeBuilder.asWriteBuilder.buildForBatch()
+        ReplaceData(target, batchWrite, mergePlan)
+    }
+  }
+
+  def buildWritePlan(childPlan: LogicalPlan,
+                     planToResolveFrom: LogicalPlan,
+                     table: Table, targetJoinAttrs: Seq[Attribute]): LogicalPlan = {
+    val iceTable = table.asInstanceOf[SparkTable].table()
+    val globalSortEnabled = isGlobalSortEnabled(table)
+    val partitionExpressions = toCatalyst(toRequiredDistribution(iceTable.spec), planToResolveFrom)
+    val sortExpressions: Seq[SortOrder] =  toCatalyst(toRequiredDistribution(iceTable.spec(),
+      iceTable.sortOrder(), false), planToResolveFrom).asInstanceOf[Seq[SortOrder]]
+    val dist = toRequiredDistribution(iceTable.spec(),
+      iceTable.sortOrder(), true)
+    val globalSortExprs: Seq[SortOrder] = toCatalyst(dist, planToResolveFrom).asInstanceOf[Seq[SortOrder]]
+
+    /*
+    val globalSortExprs: Seq[SortOrder] = toCatalyst(toRequiredDistribution(iceTable.spec(),
+      iceTable.sortOrder(), true), planToResolveFrom).asInstanceOf[Seq[SortOrder]]
+
+     */
+    val numShufflePartitions = SQLConf.get.numShufflePartitions
+
+    (globalSortEnabled, partitionExpressions.isEmpty, sortExpressions.isEmpty) match {
+      case (true, _, _) =>
+        // If global sorting is preferred then build a sort order based on partition and
+        // sort specification of the table. If none is present, then sort based on the
+        // target join key attributes.
+        if (globalSortExprs.nonEmpty) {
+          Sort(globalSortExprs, true, childPlan)
+        } else if (targetJoinAttrs.nonEmpty) {
+          Sort(buildSortOrder(targetJoinAttrs), true, childPlan)
+        } else {
+          childPlan
+        }
+      case (_, true, true) =>
+        // If no partition spec or sort order is defined for the table then repartition and
+        // locally sort the data based on the join key attributes.
+        if (targetJoinAttrs.nonEmpty) {
+          val repartition = RepartitionByExpression(targetJoinAttrs, childPlan, numShufflePartitions)
+          Sort(buildSortOrder(targetJoinAttrs), global = globalSortEnabled, repartition)
```

Comment:
Since `globalSortEnabled` is always false, it would be more clear for these cases to use `global = false`.